src/Aws/Kinesis/Client/Producer.hs

```diff
@@ -354,111 +319,99 @@ generatePartitionKey gen =
   Kin.partitionKey (T.pack name)
     & either (error ∘ T.unpack) id
 
--- | This will take up to @n@ items from a 'TBMQueue'.
+-- | This will take up to @n@ items from a 'TBMQueue', with a timeout.
 --
-takeTBMQueue
-  ∷ Int
+takeTBMQueueWithTimeout
+  ∷ ∀ α
+  . Int -- ^ number of items to get
+  → Int -- ^ timeout in microseconds
   → TBMQueue α
-  → STM [α]
-takeTBMQueue n q
-  | n <= 0 = return []
-  | otherwise = do
-      res ← tryReadTBMQueue q
-      case res of
-        Just (Just x) → (x:) <$> takeTBMQueue (n - 1) q
-        _ → return []
+  → IO [α]
+takeTBMQueueWithTimeout n timeoutDelay q = do
+  timedOutVar ← registerDelay timeoutDelay
+
+  let
+    readItems xs =
+      readTBMQueue q ≫= \case
+        Nothing → return xs
+        Just x → go (x:xs)
+
+    timeout =
+      readTVar timedOutVar ≫= check
+
+    go xs
+      | length xs ≥ n = return xs
+      | otherwise = readItems xs <|> xs <$ timeout
+
+  atomically $ go []
```

Comment:
Regarding the `thread blocked indefinitely in STM transaction` exception: you have to be careful in particular if this is latest operation in the program. This exception is triggered when a transaction is retrying but there is no other reference to any variable in the transaction. This could for instance be the case if the producer loop exists, but the consumer is not explicitly terminated. This case would be unrelated to the correctness of the inner consumer or producer loop.